File: recursion/python/solutions/take_while.py
"""Solution for take_while."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def take_while(pred, lst):
    """
    Return elements from the front while pred(x) is True.

    take_while(is_positive, 1 -> 2 -> -3 -> 4) => 1 -> 2
    take_while(is_positive, -1 -> 2 -> 3) => ()
    """
    if is_empty(lst):
        return EMPTY_LIST

    first, rest = unprepend(lst)

    return prepend(first, take_while(pred, rest)) if pred(first) else EMPTY_LIST


def is_positive(n):
    return n > 0


def is_even(n):
    return n % 2 == 0


if __name__ == '__main__':
    # Test cases
    assert to_python_list(take_while(is_positive, Node.from_list([1, 2, -3, 4]))) == [1, 2]
    assert to_python_list(take_while(is_positive, Node.from_list([-1, 2, 3]))) == []
    assert to_python_list(take_while(is_positive, EMPTY_LIST)) == []
    assert to_python_list(take_while(is_even, Node.from_list([2, 4, 5, 6]))) == [2, 4]
    assert to_python_list(take_while(is_positive, Node.from_list([1, 2, 3]))) == [1, 2, 3]

    print("All tests passed!")

File: recursion/javascript/solutions/takeWhile.js
/**
 * Solution for takeWhile.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function takeWhile(pred, list) {
  if (isEmpty(list)) {
    return EMPTY_LIST;
  }

  let [first, rest] = unprepend(list);

  if (pred(first)) {
    return prepend(first, takeWhile(pred, rest));
  }
  return EMPTY_LIST;
}

const isPositive = (n) => n > 0;

if (require.main === module) {
  console.assert(JSON.stringify(toArray(takeWhile(isPositive, Node.fromArray([1, 2, -3, 4])))) === '[1,2]');
  console.assert(JSON.stringify(toArray(takeWhile(isPositive, Node.fromArray([-1, 2, 3])))) === '[]');
  console.assert(JSON.stringify(toArray(takeWhile(isPositive, Node.fromArray([1, 2, 3])))) === '[1,2,3]');
  console.assert(JSON.stringify(toArray(takeWhile(isPositive, EMPTY_LIST))) === '[]');

  console.log('All tests passed!');
}

module.exports = { takeWhile };
